### ISSUE TITLE
Develop plan for moving update of FreeCen2 from FC1 update to direct update 

### ISSUE THREAD

**Issue by Captainkirkdawson** [FreeUKGen/FreeCENMigration#728]
Develop document that outlines basic process for the change in update process to guide the prioritization of implementation stories

E4 U6 S8 C7:P 25

**Comment by PatReynolds** [comment]
Needs discussion by @PatReynolds @richpomfret  @Captainkirkdawson and @Vino-S on actions needed to adapt multimodal ingestion of batches for FreeCEN.  Possibley on 10th Feb, after Systemwide, if all available.

**Comment by richpomfret** [comment]
@Vino-S can we look to identify key tasks involved in updating the FC DB and how we might do this directly into FC2? Do we convert VLD files into CSV? Can we use the FR tools/modify them to use a similar process? What about validation tools? etc.

**Comment by richpomfret** [comment]
@Captainkirkdawson to send @PatReynolds a list of steps/validation rules for FR.

**Comment by Captainkirkdawson** [comment]
Converted into an epic

**Comment by Captainkirkdawson** [comment]
An old document on process is contained in
https://docs.google.com/document/d/1poYSNWPjjaIH631BIHdRNnbZpw0HpGAE4ZmfzkjtEsk/edit?ts=5e41a415

**Comment by Captainkirkdawson** [comment]
@vino can you either point me to a document on the FreeCEN2 update process you currently use or write such a short document.
@lemon-ukgen can you either point me to a document on the FreeCEN1 update process you currently use or write such a short document.
@lemon-ukgen can you point me to a document on what steps you do to prepare for the update of FreeCEN2. Can you create is one does not exist.
We cannot develop a plan for a direct update until we understand the current process.

**Comment by lemon-ukgen** [comment]
The FreeCEN1 database update process -- such that it is -- is in CVS at `UKGEN-FreeCEN/README`. You can read it on vervet at `~lemon/work/UKGEN-FreeCEN/README` if you don't have a checkout handy.

@Captainkirkdawson about your second request - the only updating I have ever done on FreeCEN2 is my hand in the `update-freecen2` script to update the code. I haven't ever touched the FreeCEN2 data, so I don't think I can write such a document.

**Comment by lemon-ukgen** [comment]
Having read the linked document I think you'll find my `README` a little unsatisfactory, since it doesn't really help explain what happens with these `VLD` files.

The part of the process I've been involved in is getting the new/updated `VLD`s from Dropbox and into the FreeCEN1 filesystem so the build (via `make all`) can proceed. The backstory is that the Dropbox route came about because the previous drop-off system (Windows Network Places `DAV`) stopped working and we needed Something to replace it. Dropbox plus some scripting became that Something.

As for what that actual build *does* the best place to look is the `Makefile` and the scripts it invokes, see `vervet:~apache/hosts/freecen/bin/Makefile` and neighbours. I can't comment further, I've never needed go inside myself.

**Comment by Captainkirkdawson** [comment]
List sent of validations for FR sent to @PatReynolds

**Comment by Captainkirkdawson** [comment]
Thank you @lemon-ukgen; as you say the README is as clear as mud
After lots of digging I have come to the conclusion my understanding of CEN, its structure and processing is sufficient that any suggestions I might make are likely to be highly dangerous and likely inappropriate. 
It also seems that everything revolves around Brenda and is unlikely to function without her and her brain is the documentation on CEN offline processing. 
The input to CEN1 is apparently PARMS from Brenda to lemon and lemon taking pieces from a dropbox. These are loaded onto colobus.The input to CEN2 is the input to the CEN1; obtained by syncing three folders; fixed, pieces and db-stats. The terminology of fixed and its relationship to PARMS is unclear as is the role of db-stats
The CEN2 updating is apparently following a process documented in https://docs.google.com/document/d/1emA_3FtPGs5iOvmce4C8pESKpTZ957-1yM8KZzxz754/edit
There are however update scripts on the servers that indicate that 3 years ago the process was automated.
https://github.com/FreeUKGen/MyopicVicar/blob/master/lib/tasks/scripts/update_freecen2_monthly_update.sh
The update process for CEN2 is self documented in the code
0: Doing some consistency checks on the database data before starting update
1a) reading ctyPARMS.dat files
1b) reading .VLD validated piece files
      These are both then separated into 5 sets DELETED MULTIPLE NEW MODIFIED UNCHANGED
2. Deleting DELETED VLDs from database
3 Deleting MODIFIED VLDs from database to be reloaded
4. Deleting MULTIPLE VLDs from database to try to fix:
5. Dropping DELETED PARMS (and all associated VLDs)
6. Dropping MULTIPLE PARMS (and all associated VLDs)
6. Dropping MODIFIED PARMS (and all associated VLDs)
7. Loading/Reloading NEW  and MODIFIED PARMS files
8. Loading/reloading NEW and MODIFIED VLDs
9 Do some consistency checks on the database data

As a separate task Update piece subplaces geolocation info for PARMS that were loaded

Conceptually then it is possible to envisage CEN2 being updated off line separately and at any time by  having a process that loads a vld file from  the drop box and /or a parms into their respective folders for CEN2 except for the role played by db-stats.
Furthermore it is possible to envisage on-line processes whereby transcribers would upload vld files. Coordinators could transfer these to proofreaders /checkers. Those people could either download or online review the file and make editing changes. If done online then they could be flagged for processing into the database. There could be a nightly processing or more frequent of the "accepted" vlds. 
Similar processes could be adopted for parms although I remain confused over parms and fixed and what iis the difference.

**Comment by FreecenBren** [comment]
Hi Kirk,
Many thanks.
I was ok at the start but got confused when I hit this sentence as the VLD
is the final stage  that Lemon uploads so not sure what COORDS would
transfer?
Possible typo?

*Furthermore it is possible to envisage on-line processes whereby
transcribers would upload vld files. Coordinators could transfer these to
proofreaders /checkers*

Anyway, my penny worth for what it is worth, as no one appears to mention
the FCTOOLS which is my lifeline.

This is because the process of FreeCEN One is,
CSV, DAT, CHE and the final stage Is  the VLD .
So that is why I am confused as to how the Transcriber gets gold of a VLD?
Although in the future will we need them all?

So just a quick summary of FC1 as it has evolved over the years.

1. Transcriber uses a spreadsheet that is saved as a CSV using the PIECE
number as the unique reference point.

2.  There are two options for the Proof Reading Stage.

A. The completed CSV then gets converted to a DAT file using FCTOOLS
program, and is zipped together with the correct PARMS file for the County
x Census Year being worked on.
This zipped folder is  then sent to the Proof Reader for the next stage
using WINCC.

 B. Or Alternatively the Proof Reader might prefer the ‘Visual Checking’
method as WINCC is not always friendly on Windows 10 PC or the Proof Reader
might be using a iPad which WINCC will not work in anyway.

3.  Depending on which route was taken for the Proof Reading stage the
Validating  is then done using VALDREV.

    A) if they used WINCC, then when the Proof Reader has finished the
WINCC program will convert the DAT to a CHE and produce a ZIPPED file (this
 contains the CHE and the PARMS file)

    B) If the Proof Reader used Visual Checking method instead, then they
will forward the finished CSV back to the Coordinator.
The COORD will then convert the CSV to a CHE using FCTOOLS and will then
zip it with the appropriate PARMS file  ready for the final Validating
process.

4. It does not really matter which Proof Reading system they used, either
2A or 2B as above, the resulting Zipped folder is then copied into the
*VALDREV* program for the Validator.

5. When all the entries have been Validated  the final *VLD* is produced.

6. This is then sent to me for inclusion in the next upload for Lemon.

This then takes us back to the sentence at the start that I got confused
about as the *Transcriber* would never see the *VLD* as that is the
finished product.

Having said all that and what each stage does, there is an option that
using *FCTOOLS* it is possible that after the *CSV* has been completed by
the *TRANSCRIBER*, we can miss out WINCC and  go straight through to the
*VALDREV* stage with the *Proof Reader*  to get to the *VLD*

So I think it would be a good idea for someone to actually look at what
*FCTOOLS* does and possibly talk to Rick Parsons who designed this
wonderful program that COORDS would be completely lost without!

Just to add a note that I personally think would be a good idea is that
when online transcribing is available the accuracy of the data entered
would improve, as many errors are caused at the moment by Columns missed
out, too many digits and POB not spelt correctly to name just a few.

So if that is sorted then the Proof Reading and Validating stages could  be
joined together to produce the final VLD.
The COORD could then have a final look and click an upload button (Or
whatever)

*FCTOOLS* already does this and that then takes me back to a talk with Rick.

Anyway, these are only suggestions from a none programmer who just loves
FCTOOLS which is my lifeline.

**Comment by Captainkirkdawson** [comment]
@FreecenBren That you there were lots of very useful stuff in your notes which helps in my education if nothing more.
You mention FCTOOLS ; the extentent of mu knowledge is such that I have never heard the term before. What is it? What does it do? Who uses it? Why do they need it?.

**Comment by Captainkirkdawson** [comment]
You also mention CHE and DAT and other bits and pieces that again mean nothing to me but are clearly of great importance to the current process but I have to ask myself are they or should they be relevant to the operation of CEN2. Also could someone point me to Rick Parsons if that is appropriate.
Seems to me there is a great deal of packaging and repackaging and transmitting data around that could be eliminated.
The real question for management is whether they want me to spend time on getting the required understanding and translating that into code and processes within CEN2 or perhaps @Vino-S already knows this.

**Comment by Captainkirkdawson** [comment]
Note by Doug
https://docs.google.com/document/d/1GvZLE1MbY_WkV-ciP9lM7-JJKe4sd7vmjQjI2WQNAEA/edit?ts=5e4d74f5

Fctools
https://west-penwith.org.uk/fctools/

![cen_transcriber_file_names.JPG](https://images.zenhubusercontent.com/5b68ae684b5806bc2bdefcf0/506aa05e-0439-478a-ab72-39d6a005499e)

**Comment by Captainkirkdawson** [comment]
@FreecenBren Brenda @geoffj-FUG. I now fully understand why the following paragraph caused Brenda such confusion
_Furthermore it is possible to envisage on-line processes whereby transcribers would upload vld files. Coordinators could transfer these to proofreaders /checkers. Those people could either download or online review the file and make editing changes. If done online then they could be flagged for processing into the database. There could be a nightly processing or more frequent of the "accepted" vlds._

Would the following make sense?

It is possible to envisage a process on CEN2 whereby transcribers would upload their csv file to CEN2 when ready. They would receive a report from CEN2 somewhat along the lines of that produced by CSVCHECK. The transcriber corrects their csv file on their computer and uploads a replacement file. They again receive a report. This can be repeated until the transcriber is satisfied with the quality of their work and wishes it to go to the checking stage. They could do this work incrementally if they wished rather than waiting until the end.

They inform their coordinator when the csv file is ready for proofreading.  The coordinator assigns a proof reader to review the file. The proof reader downloads the csv file file, the last processing report and any other files they need. They following a similar process of off line editing and uploading until they are satisfied. The proof reader informs the Coordinator that the file is ready for checking. The coordinator assigns a checker who follows the same process as the proofreader  of downloading the csv file, the report and any other files they need. Thy inform the coordinator when they are satisfied with the file. At that point the Coordinator does any final validations and indicates to CEN2 that the csv file can be processed into the database.

In this process there is only a csv file. No format changing is required. Each version of the csv file and its associated report would be archived on replacement thereby maintaining a complete history of what had been done to the file.

**Comment by FreecenBren** [comment]
Hi Kirk,
Very good.

The part I have underlined at the bottom needs replacing with the following.
 There are only the 3 Stages.

1 Transcriber
2 Proof Reader (was called a Checker)
3 Validator (and uploader for want of a better word)


“*The Proof reader informs the Coordinator that the file is ready for
Validating. *
*The coordinator assigns a Validator (Often the COORD or one of their
trusted helpers that the COORD has trained)*
*The Validator follows the same process as the proofreader of downloading
the csv file, the report and any other files they need.*
*At  that point the Coordinator does any final validations and indicates to
CEN2 that the CSV file can be processed into the database.”*


Well done. You did your homework very well!
Brenda

**Comment by Captainkirkdawson** [comment]
Revised process reflecting Brenda's comments;
A transcriber receives an assignment from their coordinator and prepares a csv file using spreadsheet software following currently established practices. The transcribers uploads their csv file to CEN2 when ready. They will receive a report from CEN2 somewhat along the lines of that produced by CSVCHECK. The transcriber corrects their csv file on their computer and uploads a replacement file. They again receive a report. This can be repeated until the transcriber is satisfied with the quality of their work and wishes it to go to the next stage. ( Note they could do their transcribing and checking incrementally if they wished rather than waiting until it is all transcribed.)

They inform their coordinator when the csv file is ready for proofreading. The coordinator assigns a proof reader to review the file. The proof reader downloads the csv file file, the last processing report and any other files they need. They following a similar process of off line editing and uploading until they are satisfied. The proof reader informs the Coordinator that the file is ready for validation. The coordinator assigns a validator (Often the Coordinator themselves or one of their
trusted helpers that the Coordinator has trained) who follows the same process as the proofreader of downloading the csv file, the report and any other files they need. The validator inform the coordinator when they are satisfied with the file. At that point the Coordinator does any final validations and when satisfied indicates to CEN2 that the csv file can be processed into the database.

In this process there is only a csv file. No format changing is required. Each version of the csv file and its associated report would be archived on replacement thereby maintaining a complete history of what had been done to the file.

**Comment by Captainkirkdawson** [comment]
Transcriber, checkers and validators can upload and replace their CSV files.#121 identified.
Automated processing and checking of uploaded CSV file#116

**Comment by PatReynolds** [comment]
What fields are in the VLD which are not in the CSV?

Some things don't need transcription, some do.

Andswer to the following questions depend on answer above:

Two options for things which need transcription: choice needed: added to CSV or added on uploaded CSV?  

Two options for who does the adding: Transcriber, or someone else.

**Comment by FreecenBren** [comment]
Example and explanation below from a transcribed CSV to its final format as a VLD


Please find attached all the files through to the VLD for a PIECE from Bedford (BDF) for 1861 census.
 
1.	The ZIP is a copy showing all the columns that the VLD has and associated files.
(I have not included a CHE as this was not done using WINCC)


2.	We can put the VLD in FCTOOLS and convert it back to a CSV to do any corrections/amendments.
 
3.	We can then do the corrections/amendments and convert it back to a VLD using FCTOOLS.
 
4.	Th ‘Amended VLD’ is then sent to Lemon in a sperate folder by CHP.
That is to ensure that he knows to delete the one current ONLINE and then replace it with the AMENDED one.
If he did not do that, then it would fail as a duplicate and the amended one will not get processed.
 
5.	The 2 files  called ERD & ERC are just a summary we get that should show any errors or not  whatever the case may be etc.  Just click on them are they will open in NOTEPAD.
 
6.	The separate CSV is the ‘transcribed copy ‘ before it goes to the PROOF READER.
 
a.	The TXT is the report from the CSVCHECK program.
 
Anything else you need let me know.
 
Brenda
